src/epoch/mod.rs

```diff
@@ -135,99 +130,81 @@ mod participant;
 mod participants;
 
 pub use self::atomic::Atomic;
-pub use self::guard::{pin, Guard};
+pub use self::guard::{pin, Guard, Pinned};
 
-use std::ops::{Deref, DerefMut};
-use std::ptr;
-use std::mem;
+use std::ops;
 
-/// Like `Box<T>`: an owned, heap-allocated data value of type `T`.
-#[derive(Debug)]
-pub struct Owned<T> {
-    data: Box<T>,
+/// Like `&'a T`: a shared reference valid for lifetime `'a`.
+///
+/// This implicitly asserts that an epoch is active in this thread for its lifetime, which is what
+/// makes this different from `&'a T`.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Shared<'a, T: 'a> {
+    /// The inner reference.
+    data: &'a T,
 }
 
-impl<T> Owned<T> {
-    /// Move `t` to a new heap allocation.
-    pub fn new(t: T) -> Owned<T> {
-        Owned { data: Box::new(t) }
-    }
-
-    fn as_raw(&self) -> *mut T {
-        self.deref() as *const _ as *mut _
+impl<'a, T> Shared<'a, T> {
+    /// Create a `Shared` pointer from a reference.
+    ///
+    /// # Safety
+    ///
+    /// This is unsafe as it assumes the call side assure an epoch is valid for `'a`.
+    unsafe fn from_ptr(r: *const T) -> Shared<'a, T> {
+        Shared {
+            data: &*r,
+        }
     }
 
-    /// Move data out of the owned box, deallocating the box.
-    pub fn into_inner(self) -> T {
-        *self.data
+    /// Map the pointer to some other pointer.
+    ///
+    /// This applies `f` to the reference and returns the result wrapped in `Shared`.
+    ///
+    /// # Safety
+    ///
+    /// This is safe as it preserves the invariant that `'a` spans an epoch pin.
+    pub fn map<U: 'a, F>(self, f: F) -> Shared<'a, U>
+        where F: FnOnce(&'a T) -> &'a U {
+        unsafe { Shared::from_ptr(f(&self)) }
     }
-}
 
-impl<T> Deref for Owned<T> {
-    type Target = T;
-    fn deref(&self) -> &T {
-        &self.data
+    /// Obtain the raw pointer to the inner data.
+    ///
+    /// # Safety
+    ///
+    /// Creating a raw pointer is not unsafe, but using it is.
+    pub fn as_raw(&self) -> *mut T {
+        self.data as *const _ as *mut _
     }
-}
 
-impl<T> DerefMut for Owned<T> {
-    fn deref_mut(&mut self) -> &mut T {
-        &mut self.data
+    /// Convert it to a owned guard/pinned reference.
+    ///
+    /// This converts `self` into a value holding both `guard` and the reference, getting rid of
+    /// the bound epoch guard by instead storing it (owned) in the return value.
+    pub fn into_pinned(self, guard: Guard) -> Pinned<&'a T> {
```

Comment:
> // Using another Guard here is unsound, since the Shared we obtained is only bound to `grd`.

Not true. The `Shared` type only asserts that a guard is active.